File: src/simulation/endogenous-events.js
import { activateCityDetail, activateTerritoryDetail, locateEventTerritory } from "./grid-spatial.js";
import { recordEvent } from "./journal.js";

export function advanceEndogenousEvents(world) {
  if (world.day === 0 || world.day % 30 !== 0) return;
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    if (Object.keys(city.activeEffects).length > 0) continue;
    const node = world.spatial.nodes[city.spatialNodeId];
    const wetlandShare = (node.aggregate.biomeShares.wetland ?? 0) +
      (node.aggregate.biomeShares.floodplain ?? 0) * 0.35;
    const density = node.aggregate.population / Math.max(1, node.childTerritoryIds.length);
    const traffic = world.journal.filter((event) =>
      event.day >= world.day - 30 && event.type === "shipment_arrived" && event.details.to === cityId
    ).length;
    const risk = 0.0008 + wetlandShare * 0.003 + Math.min(0.0015, density / 25_000) +
      (1 - city.demography.health) * 0.006 + Math.min(0.001, traffic * 0.00005);
    if (world.randomStreams.events.next() >= risk) continue;

    const durationDays = 24 + Math.floor(world.randomStreams.events.next() * 43);
    const multiplier = 0.58 + world.randomStreams.events.next() * 0.2;
    const territoryId = locateEventTerritory(world, cityId, world.randomStreams.events);
    const recentArrivalCauses = world.journal
      .filter((event) => event.day >= world.day - 30 && event.type === "shipment_arrived" && event.details.to === cityId)
      .slice(-3)
      .map((event) => event.id);
    const effectId = `endemic-fever:${cityId}:${world.day}`;
    const startEvent = recordEvent(world, {
      type: "crisis_started",
      subjectId: effectId,
      causeIds: recentArrivalCauses,
      details: {
        cityId,
        territoryId,
        label: `Лихорадка в поселении «${city.name}»`,
        multiplier,
        durationDays,
        endogenous: true,
        emergence: { wetlandShare, density, health: city.demography.health, traffic, risk }
      }
    });
    city.activeEffects[effectId] = {
      multiplier,
      endDay: world.day + durationDays,
      startEventId: startEvent.id,
      territoryId,
      label: startEvent.details.label,
      endogenous: true
    };
    activateCityDetail(world, cityId, {
      causeEventId: startEvent.id,
      keepActiveDays: durationDays + world.lodPolicy.crisisCooldownDays
    });
    activateTerritoryDetail(world, territoryId, {
      causeEventId: startEvent.id,
      keepActiveDays: durationDays + world.lodPolicy.crisisCooldownDays
    });
  }
}

File: src/WorldGen.Core/Simulation/EndogenousEvents.cs
using System.Text.Json.Nodes;

namespace WorldGen.Core.Simulation;

public static class EndogenousEvents
{
    public static void Advance(WorldState world)
    {
        if (world.Day == 0 || world.Day % 30 != 0) return;
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId]; if (city.ActiveEffects.Count > 0) continue;
            var node = world.Spatial.Nodes[city.SpatialNodeId];
            var wetland = node.Aggregate.BiomeShares.GetValueOrDefault("wetland") + node.Aggregate.BiomeShares.GetValueOrDefault("floodplain") * 0.35;
            var density = node.Aggregate.Population / (double)Math.Max(1, node.ChildTerritoryIds!.Count);
            var arrivals = world.Journal.Where(evt => evt.Day >= world.Day - 30 && evt.Type == "shipment_arrived" &&
                evt.Details["to"]?.GetValue<string>() == cityId).ToArray();
            var traffic = arrivals.Length;
            var risk = 0.0008 + wetland * 0.003 + Math.Min(0.0015, density / 25_000) +
                (1 - city.Demography.Health) * 0.006 + Math.Min(0.001, traffic * 0.00005);
            var random = world.RandomStreams["events"];
            if (random.NextDouble() >= risk) continue;
            var duration = 24 + (int)Math.Floor(random.NextDouble() * 43);
            var multiplier = 0.58 + random.NextDouble() * 0.2;
            var territoryId = SpatialRuntime.LocateEventTerritory(world, cityId, random);
            var causes = arrivals.TakeLast(3).Select(evt => evt.Id);
            var effectId = $"endemic-fever:{cityId}:{world.Day}";
            var label = $"Лихорадка в поселении «{city.Name}»";
            var evt = Journal.Record(world, "crisis_started", effectId, causes,
                new JsonObject { ["cityId"] = cityId, ["territoryId"] = territoryId, ["label"] = label,
                    ["multiplier"] = multiplier, ["durationDays"] = duration, ["endogenous"] = true,
                    ["emergence"] = new JsonObject { ["wetlandShare"] = wetland, ["density"] = density,
                        ["health"] = city.Demography.Health, ["traffic"] = traffic, ["risk"] = risk } });
            city.ActiveEffects[effectId] = new ActiveEffectState(multiplier, world.Day + duration, evt.Id, territoryId, label, true);
            SpatialRuntime.ActivateCityDetail(world, cityId, evt.Id, duration + world.LodPolicy.CrisisCooldownDays);
            SpatialRuntime.ActivateTerritoryDetail(world, territoryId, evt.Id, duration + world.LodPolicy.CrisisCooldownDays);
        }
    }

    public static void EndExpiredEffects(WorldState world)
    {
        foreach (var city in world.Cities.Values)
        foreach (var pair in city.ActiveEffects.OrderBy(pair => pair.Key, StringComparer.Ordinal).ToArray())
        {
            if (pair.Value.EndDay != world.Day) continue;
            Journal.Record(world, "crisis_ended", pair.Key, [pair.Value.StartEventId],
                new JsonObject { ["cityId"] = city.Id, ["label"] = pair.Value.Label, ["endogenous"] = pair.Value.Endogenous });
            city.ActiveEffects.Remove(pair.Key);
        }
    }
}
